File: 0623-add-one-row-to-tree/0623-add-one-row-to-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} val
 * @param {number} depth
 * @return {TreeNode}
 */
var addOneRow = function (root, val, depth) {
    if (depth==1){
        const node = new TreeNode(val);
        node.left= root;
        return node

    }

    var level = 1;
    const Queue = []
    Queue.push(root)
    let flag = false;

    while (Queue.length > 0) {
        let QueueLength = Queue.length;
        for (let i = 0; i < QueueLength; i++) {
            if (level == depth - 1) {
                flag = true;
                let tempNode = Queue[0];
                Queue.shift();
                let tempNodeleft = tempNode.left;
                let tempNoderight = tempNode.right;
                let l1 = new TreeNode(val);
                let l2 = new TreeNode(val);
                tempNode.left = l1;
                tempNode.right = l2;
                if (tempNodeleft) {
                    l1.left = tempNodeleft

                }
                if (tempNoderight) {
                    l2.right = tempNoderight

                }


            }
            else{


            let temp = Queue[0];


            Queue.shift()
            if (temp.left != null) {
                Queue.push(temp.left)
            }
            if (temp.right != null) {
                Queue.push(temp.right)

            }
            }
        }
        if (flag == true){
            return root
        }
        level++;


    }
    return root;


};

File: 0623-add-one-row-to-tree/0623-add-one-row-to-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
         if(depth==1){
            TreeNode* newroot = new TreeNode(val);
            newroot->left= root;
            return newroot;
         }
         bool rowadded= false;
        int currentdepth = 1;
        queue<TreeNode*> q;
        q.push(root);
       

        while (q.size() > 0 && !rowadded) {
            int qsize = q.size();

            for (int i = 0; i < qsize; i++) {
                TreeNode* temp = q.front();
                q.pop();
                if (currentdepth == depth - 1) {
                    TreeNode* leftNode = temp->left == NULL ? NULL : temp->left;
                    TreeNode* rightNode =
                        temp->right == NULL ? NULL : temp->right;
                    TreeNode* first = new TreeNode(val);
                    TreeNode* second = new TreeNode(val);
                    temp->left = first;
                    temp->right = second;
                    first->left = leftNode;
                    second->right = rightNode;
                    rowadded= true;
                } else {
                    if (temp->left) {
                        q.push(temp->left);
                    }
                    if (temp->right) {
                        q.push(temp->right);
                    }
                }
            }

            currentdepth++;
        }
        return root;
    }
};